### PR TITLE
Allow votes for old headers up to gc_depth

### PR DESCRIPTION
**Pull request by aschran**

### CODE REVIEW

**Comment by mwtian** [comment]
More comment on the background?

**Comment by aschran** [comment]
Background: It's not incorrect to vote on  old headers as long as we are not equivocating, theoretically a slow node could catch up and its chain of old certs could be included in a later commit. At the moment the new refactored code is too restrictive on how old of a header it will vote on and rejects votes for headers that are e.g. even 1 round old.

**Review by mwtian — COMMENTED** [review]
I see. Is there a real problem that has been observed? The issue is that to include the past certificates (max round R) in a new proposed header / certificate and ensure they get included in the DAG, IIUC there are additional conditions:
1. If a node receives a certificate at round > R + 1, the node would try to propose at the higher round instead. So certificates at round R would not be included as parents in the proposed header.
2. When a node produces a certificate at round R, other nodes that have already advanced to round > R + 1 cannot include this certificate in parents.

My intuition is that the above two conditions are a bit unlikely, unless a node intentionally tries to create a long "branch" of certificates in the dag.

Alternatively, I think the current logic can be relaxed just to
```
ensure!(narwhal_round - 1 <= header.round, ...);
```
So if the voted header is able to form a certificate, the new certificate still has a chance to be included as parent in the voting node. What do you think?

File: narwhal/primary/src/primary.rs [hunk 1]
```diff
@@ -684,9 +682,9 @@ impl PrimaryReceiverHandler {
 
         // Now that we've got all the required certificates, ensure we're voting on a
         // current Header.
-        narwhal_round = *rx_narwhal_round_updates.borrow();
+        let narwhal_round = *rx_narwhal_round_updates.borrow();
         ensure!(
-            narwhal_round <= header.round,
+            narwhal_round.saturating_sub(self.gc_depth) <= header.round,
```
Comment (mwtian) [inline narwhal/primary/src/primary.rs:687]: The actual GC round is `commit round - gc depth`, which is usually lower than `narwhal round - gc depth`. As discussed, we don't have to re-use the gc depth value. A separate constant might be better?

**Comment by aschran** [comment]
This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?

**Comment by mwtian** [comment]
> This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?

I agree the downside of more permissive voting is not that significant. My only remaining issue is that can we avoid the `narwhal_round.saturating_sub(self.gc_depth)` calculation because I think the result is not meaningful? Either pass in `rx_consensus_round_updates` and calculate the GC round as `rx_consensus_round_updates.borrow() - gc_depth`, or as I mentioned use another constant to subtract from `narwhal_round` and avoid passing in `gc_depth`?

**Comment by aschran** [comment]
> > This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?
> 
> I agree the downside of more permissive voting is not that significant. My only remaining issue is that can we avoid the `narwhal_round.saturating_sub(self.gc_depth)` calculation because I think the result is not meaningful? Either pass in `rx_consensus_round_updates` and calculate the GC round as `rx_consensus_round_updates.borrow() - gc_depth`, or as I mentioned use another constant to subtract from `narwhal_round` and avoid passing in `gc_depth`?

Can you explain more what do you mean by the result is not meaningful? Narwhal round here is more relevant than consensus round because this header signing is part of the narwhal algorithm and not directly relevant to the consensus algorithm. We could add a second gc_depth parameter but that seemed excessive to me (just unnecessary extra code) when we have no desire at this time for gc_depth to differ?

**Comment by mwtian** [comment]
> > > This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?
> > 
> > 
> > I agree the downside of more permissive voting is not that significant. My only remaining issue is that can we avoid the `narwhal_round.saturating_sub(self.gc_depth)` calculation because I think the result is not meaningful? Either pass in `rx_consensus_round_updates` and calculate the GC round as `rx_consensus_round_updates.borrow() - gc_depth`, or as I mentioned use another constant to subtract from `narwhal_round` and avoid passing in `gc_depth`?
> 
> Can you explain more what do you mean by the result is not meaningful? Narwhal round here is more relevant than consensus round because this header signing is part of the narwhal algorithm and not directly relevant to the consensus algorithm. We could add a second gc_depth parameter but that seemed excessive to me (just unnecessary extra code) when we have no desire at this time for gc_depth to differ?

I mean `narwhal proposing round - gc depth` means nothing elsewhere in the narwhal / consensus logic. GC round is calculated by `committed round - gc depth`. Again, I would much prefer using a different named constant if you want to use `narwhal round`, or keep the previous gc round calculation.

**Comment by aschran** [comment]
> > > > This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?
> > > 
> > > 
> > > I agree the downside of more permissive voting is not that significant. My only remaining issue is that can we avoid the `narwhal_round.saturating_sub(self.gc_depth)` calculation because I think the result is not meaningful? Either pass in `rx_consensus_round_updates` and calculate the GC round as `rx_consensus_round_updates.borrow() - gc_depth`, or as I mentioned use another constant to subtract from `narwhal_round` and avoid passing in `gc_depth`?
> > 
> > 
> > Can you explain more what do you mean by the result is not meaningful? Narwhal round here is more relevant than consensus round because this header signing is part of the narwhal algorithm and not directly relevant to the consensus algorithm. We could add a second gc_depth parameter but that seemed excessive to me (just unnecessary extra code) when we have no desire at this time for gc_depth to differ?
> 
> I mean `narwhal proposing round - gc depth` means nothing elsewhere in the narwhal / consensus logic. To me, it is kind of like comparing apple to oranges ... GC round is calculated by `consensus committed round - gc depth`. I would much prefer using a different named constant if you want to use `narwhal round`, then you would be able to avoid passing in the `gc depth` parameter and make future adjustments easier. Or keeping the previous gc round calculation is fine too.

Does it not have roughly the same meaning? We are garbage collecting old vote requests beyond the gc_depth threshold. It's just computing it from the narwhal round, because voting is a narwhal concept, instead of consensus round (which is not relevant to narwhal voting). It's not really that different from your use of gc_round in the CertificateWaiter, which is also not part of consensus round? (CertificateWaiter should probably **also** use Narwhal round and not consensus round, since it is used to run the narwhal algorithm and not consensus.)

**Comment by mwtian** [comment]
> > > > > This just restores previous behavior from before the refactor, based on a long Slack discussion with George, Alberto, Lefteris. I don't see a big downside to voting on some certs that maybe don't end up in the DAG? The current behavior is in any case definitely too constraining. I think if we want to consider a third option that is neither previous behavior nor current, maybe it can be in a follow-up PR?
> > > > 
> > > > 
> > > > I agree the downside of more permissive voting is not that significant. My only remaining issue is that can we avoid the `narwhal_round.saturating_sub(self.gc_depth)` calculation because I think the result is not meaningful? Either pass in `rx_consensus_round_updates` and calculate the GC round as `rx_consensus_round_updates.borrow() - gc_depth`, or as I mentioned use another constant to subtract from `narwhal_round` and avoid passing in `gc_depth`?
> > > 
> > > 
> > > Can you explain more what do you mean by the result is not meaningful? Narwhal round here is more relevant than consensus round because this header signing is part of the narwhal algorithm and not directly relevant to the consensus algorithm. We could add a second gc_depth parameter but that seemed excessive to me (just unnecessary extra code) when we have no desire at this time for gc_depth to differ?
> > 
> > 
> > I mean `narwhal proposing round - gc depth` means nothing elsewhere in the narwhal / consensus logic. To me, it is kind of like comparing apple to oranges ... GC round is calculated by `consensus committed round - gc depth`. I would much prefer using a different named constant if you want to use `narwhal round`, then you would be able to avoid passing in the `gc depth` parameter and make future adjustments easier. Or keeping the previous gc round calculation is fine too.
> 
> Does it not have roughly the same meaning? We are garbage collecting old vote requests beyond the gc_depth threshold. It's just computing it from the narwhal round, because voting is a narwhal concept, instead of consensus round (which is not relevant to narwhal voting). It's not really that different from your use of gc_round in the CertificateWaiter, which is also not part of consensus round? (CertificateWaiter should probably **also** use Narwhal round and not consensus round, since it is used to run the narwhal algorithm and not consensus.)

As discussed off thread, GC round / depth is a concept that flows from consensus DAG construction / output. Narwhal drops data that will not contribute to DAG (at or below GC round). But semantically it cannot be switched with `narwhal proposing round - gc depth`, which can advance arbitrarily ahead of committed round. Using `narwhal proposing round - gc depth` is very weird in my opinion, since you have to pass the extra `gc depth` parameter. But other folks may think differently.

**Comment by aschran** [comment]
Per Slacks with George + Mingwei, updated this to still use narwhal round with a constant age limit.

**Review by mwtian — APPROVED** [review]

**Pull request merged**